src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -443,6 +449,45 @@ protected void handshakeFailure(final ChannelHandlerContext context, final Throw
     }
 
     /**
+     * Returns socket channel class suitable for specified event loop group.
+     *
+     * @param eventLoopGroup
+     * @return socket channel class
+     * @throws IllegalArgumentException in case of null or unrecognized event loop group.
+     */
+    private Class<? extends Channel> getSocketChannelClass(EventLoopGroup eventLoopGroup) {
+        if (eventLoopGroup == null) {
+            log.warn("Asked for socket channel class to work with null event loop group, returning NioSocketChannel class.");
+            return NioSocketChannel.class;
+        }
+
+        if (eventLoopGroup instanceof NioEventLoopGroup) {
+            return NioSocketChannel.class;
+        } else if (eventLoopGroup instanceof OioEventLoopGroup) {
+            return OioSocketChannel.class;
+        }
+
+        // epoll?
+        String className = eventLoopGroup.getClass().getName();
+        if (EPOLL_EVENT_LOOP_GROUP_CLASS.equals(className)) {
```

Comment:
Only EpollEventLoopGroup lives in a separate java artifact (netty-native-epoll-linux-<arch>). We're doing instanceof here because it's cheap to code and because NioXXX and OioXXX classes are always available.

I've implemented something similar also for netty, see https://github.com/netty/netty/pull/5050, because pushy is like third library that needs this kind of patch in order to support epoll on linux.